folding-schemes/src/folding/hypernova/circuits.rs

```diff
@@ -700,31 +714,88 @@ where
         U_i1.C = U_i1_C;
 
         // P.4.a compute and check the first output of F'
-        // Base case: u_{i+1}.x[0] == H((1, z_0, z_{i+1}, U_{i+1})
-        // Non-base case: u_{i+1}.x[0] == H((i+1, z_0, z_{i+1}, U_{i+1})
         let (u_i1_x, _) = U_i1.clone().hash(
             &crh_params,
             i + FpVar::<CF1<C1>>::one(),
             z_0.clone(),
             z_i1.clone(),
         )?;
-        let (u_i1_x_base, _) = U_i1.hash(
-            &crh_params,
-            FpVar::<CF1<C1>>::one(),
-            z_0.clone(),
-            z_i1.clone(),
+        let x = FpVar::new_input(cs.clone(), || Ok(self.x.unwrap_or(C1::ScalarField::zero())))?;
+        x.enforce_equal(&u_i1_x)?;
+
+        // convert rho_bits to a `NonNativeFieldVar`
+        let rho_nonnat = {
+            let mut bits = rho_bits;
+            bits.resize(C1::BaseField::MODULUS_BIT_SIZE as usize, Boolean::FALSE);
+            NonNativeUintVar::from(&bits)
+        };
+
+        // CycleFold part
+        // C.1. Compute cf1_u_i.x and cf2_u_i.x
+        let cf_x = vec![
+            rho_nonnat, U_i.C.x, U_i.C.y, u_i.C.x, u_i.C.y, U_i1.C.x, U_i1.C.y,
+        ];
+
+        // ensure that cf_u has as public inputs the C from main instances U_i, u_i, U_i+1
+        // coordinates of the commitments.
+        // C.2. Construct `cf_u_i`
+        let cf_u_i = CycleFoldCommittedInstanceVar {
+            // cf1_u_i.cmE = 0
+            cmE: GC2::zero(),
```

Comment:
I haven't seen a check on this. Are we assuming that since it is a constant it does not need to be checked?
(Same question for the initialization of `u` below)

Reply:
since it is hardcoded to be `GC2::zero()`, it is zero, no?

Reply:
I was confused by [this comment](https://github.com/privacy-scaling-explorations/sonobe/blob/fd942bda713082239738c97c02a1e8ad02ecdfd5/folding-schemes/src/folding/nova/cyclefold.rs#L190). Since the protocol does mention that the verifier should reject if `cmE` is not zero or if `u` is not one, maybe adding this check explicitly in the circuit could be worth it?

Reply:
hmm but here the 'verifier' is run in the circuit, and the circuit has hardcoded the value to be zero, which should be ~equivalent to enforcing it / checking it to be zero, no?

Reply:
(like, if the prover has used a non-zero value, the circuit verification (proof verification) will not pass)

Reply:
Yes, that makes sense, the prover will not be able to manipulate those constants. Maybe adding a comment about this to make your logic explicit could be helpful (since the check is explicitly mentioned in the protocol but implicit here).
```suggestion
            cmE: GC2::zero(), // we ensure that the point is zero by allocating it as a circuit constant
```

Reply:
Ok, added a comment explaining that when setting the `cmE: GC2::zero()`, we don't check that it is zero in the circuit because it is already allocated as zero.